tfx/experimental/pipeline_testing/stub_component_launcher.py

```diff
@@ -0,0 +1,81 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""stub component launcher which launches component executors and stub executors in process."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+from typing import Any, Dict, List, Text, Type
+
+from tfx import types
+from tfx.components.base import base_executor
+from tfx.experimental.pipeline_testing import base_stub_executor
+from tfx.orchestration.launcher import in_process_component_launcher
+
+class StubComponentLauncher(
+    in_process_component_launcher.InProcessComponentLauncher):
+  """Responsible for launching a stub executor.
+
+  The executor will be launched in the same process of the rest of the
+  component, i.e. its driver and publisher.
+  """
+
+  def _run_executor(self, execution_id: int,
+                    input_dict: Dict[Text, List[types.Artifact]],
+                    output_dict: Dict[Text, List[types.Artifact]],
+                    exec_properties: Dict[Text, Any]) -> None:
+    """Execute underlying component implementation."""
+    component_id = self._component_info.component_id
+    if component_id not in self.component_map:
+      super(StubComponentLauncher, self)._run_executor(execution_id,
+                                                       input_dict,
+                                                       output_dict,
+                                                       exec_properties)
+    else:
+      executor_context = base_executor.BaseExecutor.Context(
+          beam_pipeline_args=self._beam_pipeline_args,
+          tmp_dir=os.path.join(self._pipeline_info.pipeline_root, '.temp', ''),
+          unique_id=str(execution_id))
+      executor = self.component_map[component_id](component_id,
+                                                  self.test_data_dir,
+                                                  executor_context)
+      executor.Do(input_dict, output_dict, exec_properties)
+
+def create_stub_launcher_class(
+    test_data_dir: Text,
+    component_ids: List[Text],
+    component_map: Dict[Text, Type[base_stub_executor.BaseStubExecutor]]
+    ) -> Type[StubComponentLauncher]:
+  """Creates a StubComponentLauncher class
+
+  Args:
+    test_data_dir: The directory where pipeline outputs are recorded
+      (pipeline_recorder.py).
+    component_ids: List of component ids that should be replaced
+      with a BaseStubExecutor.
+    component_map: Dictionary holding user-defined stub executor.
+      These user-defined stub executors must inherit from
+      base_stub_executor.BaseStubExecutor.
+  Returns:
+    StubComponentLauncher class holding component_map with stub executors.
+  """
+  cls = StubComponentLauncher
+  cls.component_map = dict(component_map)
```

Comment:
nit: It seems like this dict all have the same value? If so this is can can be replaced with a set like `stubbed_components`

Reply:
This dict holds custom component classes, and we don't assume that user will set components with a same custom component.

Reply:
Could you add your explanation as a comment?